packages/duf/src/evo/data_converters/duf/importer/utils.py

```diff
@@ -0,0 +1,444 @@
+    return table_klass(**data_client.save_table(vertices_table)), bounding_box_go
+
+
+def indices_array_to_go(data_client, indices_array, table_klass):
+    width = indices_array.shape[1]
+    indices_schema = pa.schema([pa.field(f"n{i}", pa.uint64()) for i in range(width)])
+    indices_table = pa.Table.from_arrays(
+        [pa.array(indices_array[:, i], type=pa.uint64()) for i in range(width)],
+        schema=indices_schema,
+    )
+    return table_klass(**data_client.save_table(indices_table))
+
+
+def parts_to_go(
+    data_client, parts: dict[str, int | dict[AttributeSpec, list]], parts_klass, chunks_klass=IndexArray2_V1_0_1
+):
+    if parts:
+        parts_schema = pa.schema([pa.field("offset", pa.uint64()), pa.field("count", pa.uint64())])
+        parts_table = pa.Table.from_arrays(
+            [pa.array(parts["offset"], type=pa.uint64()), pa.array(parts["count"], type=pa.uint64())],
+            schema=parts_schema,
+        )
+        chunks_go = chunks_klass(**data_client.save_table(parts_table))
+
+        if attributes := parts["attributes"]:
+            part_attributes_go = [spec.to_go(data_client, values) for spec, values in attributes.items()]
+        else:
+            part_attributes_go = None
+
+        return parts_klass(
+            chunks=chunks_go,
+            attributes=part_attributes_go,
+        )
+    return None
+
+
+def obj_list_and_indices_to_arrays(obj_list: list[dw.BaseEntity], indices_arrays: list[NDArray]):
+    orig_num_vertices = sum(obj.VertexList.Count for obj in obj_list)
+    num_parts = len(obj_list)
+
+    layers = {obj.Layer for obj in obj_list}
+    assert len(layers) == 1, "Objects must be from the same layer to combine"
+    layer = layers.pop()
+
+    axes = ("X", "Y", "Z")
+    vertices_array = np.fromiter(
+        (getattr(vert, axis) for polyface in obj_list for vert in polyface.VertexList for axis in axes),
+        dtype=np.float64,
+        count=orig_num_vertices * 3,
+    ).reshape(orig_num_vertices, 3)
+
+    new_vertices_array, orig_to_unique = np.unique(vertices_array, return_inverse=True, axis=0)  # Ensure unique
+    if len(new_vertices_array) == orig_num_vertices:
+        # No duplicates
+        orig_to_unique = None
+    else:
+        vertices_array = new_vertices_array
+
+    attribute_specs = AttributeSpec.layer_attributes(layer)
+    attribute_names = {spec.name for spec in attribute_specs}
+    if num_parts > 1 or attribute_specs:
+        # We use parts to store object-level attributes, so we need at least a single part if we have any
+        parts = {"offset": [], "count": [], "attributes": defaultdict(list)}
+        attributes = parts["attributes"]
+
+        offset = 0
+        vertex_offset = 0
+        for obj, obj_indices_array in zip(obj_list, indices_arrays):
+            obj_num_vertices = obj.VertexList.Count
+            obj_count = len(obj_indices_array)
+
+            obj_indices_array += vertex_offset  # Shift indices to the combined vertices array

```

Comment:
This is a side-effect.  If it's by design then document it at the top of the function.